xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-default/src/main/java/org/xwiki/notifications/filters/migration/R140401000XWIKI15460DataMigration.java

```diff
@@ -201,17 +221,38 @@ private void identifyRemovedUsers(Map<String, Boolean> usersStatus,
         DefaultNotificationFilterPreference filterPreference) throws DataMigrationException
     {
         String owner = filterPreference.getOwner();
+        WikiReference currentWiki = this.getXWikiContext().getWikiReference();
 
         // Store the potential exception thrown inside computeIfAbsent, since assigning to a variable is not allowed.
         AtomicReference<Exception> exception = new AtomicReference<>();
         usersStatus.computeIfAbsent(owner, key -> {
             DocumentReference entityReference = this.resolver.resolve(key);
-            UserReference userReference =
-                this.documentReferenceUserReferenceResolver.resolve(entityReference);
-
+            WikiReference wikiReference = entityReference.getWikiReference();
             try {
-                return this.userManager.exists(userReference);
-            } catch (UserException e) {
+                // if we're on same wiki we check user presence by using the UserManager since it's the most
+                // reliable solution, and we benefit from cache.
+                if (wikiReference.equals(currentWiki)) {
+                    UserReference userReference =
+                        this.documentReferenceUserReferenceResolver.resolve(entityReference);
+                    return this.userManager.exists(userReference);
+                // if we're not and the wiki doesn't exist anymore we immediately know the user doesn't exist
+                } else if (!this.wikiDescriptorManager.exists(wikiReference.getName())) {
+                    return false;
+                // if the wiki still exist we cannot really use UserManager because other migrations might not have
+                // been applied yet (see: XWIKI-20184) so we instead rely on a low-level SQL query to check existence
+                // of the user
+                } else {
+                    String serializedName = this.entityReferenceSerializer.serialize(entityReference);
+
+                    List<Long> result = this.queryManager.createQuery(XWQL_QUERY, Query.XWQL)
```

Comment:
Using XWQL might cause the exact same issue: it will need to load the class document. I'm afraid you will have to use HQL.

Reply:
Right, I forgot this one. Thanks will change.